### PR TITLE
Self Check Question Card

### PR DESCRIPTION
**Pull request by ChinemeremChigbo**
## Notion ticket link

[Self Check Question Card](https://www.notion.so/uwblueprintexecs/Self-Check-Question-Card-5626df49a24e4dd9af204e43c03e580c)

## Implementation description
* To see slides please visit: http://localhost:3000/selfCheckQuestions.
* Please note that this is not supposed to be a functional prototype, but rather just an interactive mockup. Database connection with the buttons is yet to be added.

## Steps to test
1. Ensure that the code written is up to your standard. If it is not, please provide me with ideas / steps of how to make it better

## What should reviewers focus on?
* Once again, to see slides please visit: http://localhost:3000/selfCheckQuestions.

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: pages/selfCheckQuestions.tsx [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    ModalFooter,
+    ModalCloseButton,
+    useDisclosure,
+    Flex,
+    SimpleGrid,
+    Button,
+    ButtonGroup,
+} from "@chakra-ui/react";
+import { Header, SelfCheckQuestionCard, Footer } from "@components";
+const questionList = [
```
Comment (mkanoria) [inline pages/selfCheckQuestions.tsx:19]: We should try to remove this maybe? Can't see this data on the frontend either

File: pages/selfCheckQuestions.tsx [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        type: "slider",
+        question: "autem sunt eiusdolores nesciunt impedit?",
+        options: ["1", "2", "3", "4", "5"],
+
+        questionNumber: 5,
+    },
+];
+
+const Home: React.FC = () => {
+    const [list, setList] = React.useState(questionList);
```
Comment (mkanoria) [inline pages/selfCheckQuestions.tsx:59]:
```suggestion
    const [questionList, setQuestionList] = useState([]);
```

1. Let's just try and use `useState` instead
2. Clearer variable name with `questionList`

File: pages/selfCheckQuestions.tsx [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    const [count, setCount] = React.useState(6);
+    function changeQuestionType(id, target) {
```
Comment (mkanoria) [inline pages/selfCheckQuestions.tsx:61]:
```suggestion
    const changeQuestionType = (id, target) => {
```
Let's also use ES6 syntax
Same comment for the functions below

File: pages/selfCheckQuestions.tsx [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+                    <Button
+                        _hover={{ bg: "#121310" }}
+                        _active={{
+                            transform: "scale(0.95)",
+                        }}
+                        minWidth={"90"}
+                        color="white"
+                        background="black"
+                        variant="outline"
+                        onClick={() => {
```
Comment (mkanoria) [inline pages/selfCheckQuestions.tsx:233]: This is okay for now, let's keep in mind to update this when linking APIs

File: pages/selfCheckQuestions.tsx [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+                            variant="outline"
+                            colorScheme="black"
+                            mr={3}
+                            w={100}
+                            onClick={onClose}
+                        >
+                            Cancel
+                        </Button>
+                        <Button
+                            onClick={() => {
```
Comment (mkanoria) [inline pages/selfCheckQuestions.tsx:267]: Might have to updated when integrating this change into the actual Tool builder page

File: pages/selfCheckQuestions.tsx [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+                            colorScheme="red"
+                        >
+                            Discard
+                        </Button>
+                    </ModalFooter>
+                </ModalContent>
+            </Modal>
+            <SimpleGrid columns={1} spacing={0} px={10} py={10}>
+                <Button
+                    id={"shmexy"}
```
Comment (mkanoria) [inline pages/selfCheckQuestions.tsx:281]: "shmexy" LOL

File: public/selfCheckQuestions.json [hunk 7]
```diff
@@ -0,0 +1,47 @@
+{
```
Comment (mkanoria) [inline public/selfCheckQuestions.json:1]: Are we still using this file?

**Review by mkanoria — CHANGES_REQUESTED** [review]
The UI looks great! I left a few comments, will go through more of the PR soon

Small bug(?) but the console log on save does not show me the correct user input for any title/question values I added. I can see the number of questions and type updating, but the actual content remains the same always

File: src/components/selfCheckQuestion/index.tsx [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+                    <Menu>
+                        <Heading
+                            fontSize={20}
+                            fontWeight="500"
+                            alignSelf="center"
+                            mr={6}
+                        >
+                            {questionIndex + 1}.
+                        </Heading>
+                        <Input
```
Comment (mkanoria) [inline src/components/selfCheckQuestion/index.tsx:113]: Doesn't seem like this input has an `onChange`

File: src/components/selfCheckQuestion/index.tsx [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+                                            children={
+                                                <Circle
+                                                    size="15px"
+                                                    color="white"
+                                                    borderWidth="2px"
+                                                    borderColor="black"
+                                                ></Circle>
+                                            }
+                                        />
+                                        <Input
```
Comment (mkanoria) [inline src/components/selfCheckQuestion/index.tsx:179]: Same with this input, you probably want to add an `onChange` to this input

File: public/selfCheckQuestions.json (hunk 7 above)
Comment (ChinemeremChigbo) [reply public/selfCheckQuestions.json]: Not anymore 👍

File: pages/selfCheckQuestions.tsx [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+        newSliderRange0.push(i);
+    }
+    const [sliderRange, setSliderRange] = React.useState(newSliderRange0);
+    function changeQuestionType(id, target) {
+        const newList = list.slice(0);
+        newList[newList.findIndex((e) => e._id === id)].type = target;
+
+        setList(newList);
+    }
+    function changeAlphanumeric(id, target) {
```
Comment (nimwijetunga) [inline pages/selfCheckQuestions.tsx:78]: whats the point of this? I noticed in the UI that it didn't really change anything, am i missing something?

File: pages/selfCheckQuestions.tsx [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+
+    function removeOneOption(id, index) {
+        const newList = list.slice(0);
+        newList[newList.findIndex((e) => e._id === id)].options.splice(
+            index,
+            1,
+        );
+        setList(newList);
+    }
+    function changeSliderOption(id, lowerBound, upperBound) {
```
Comment (nimwijetunga) [inline pages/selfCheckQuestions.tsx:104]:
```suggestion
    function changeSliderRange(id, lowerBound, upperBound) {
```
To make this function name more clear

File: pages/selfCheckQuestions.tsx [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+            newSliderRange.push(i);
+        }
+
+        newList[newList.findIndex((e) => e._id === id)].options = optionList;
+        setSliderRange(newSliderRange);
+        setList(newList);
+    }
+    function moveQuesitonDown(index) {
+        const newList = list.slice(0);
+        const tempQuestion = newList[index];
```
Comment (nimwijetunga) [inline pages/selfCheckQuestions.tsx:122]:
```suggestion
        newList[index] = newList.splice(index + 1, 1, newList[index])[0]
```
fancier way of doing the swap but you can leave this as is if it's easier to read

File: pages/selfCheckQuestions.tsx [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+        newList[index] = newList[index + 1];
+        newList[index + 1] = tempQuestion;
+        setList(newList);
+    }
+    function moveQuesitonUp(index) {
+        const newList = list.slice(0);
```
Comment (nimwijetunga) [inline pages/selfCheckQuestions.tsx:128]:
```suggestion
        newList[index] = newList.splice(index - 1, 1, newList[index])[0]
```
fancier way of doing the swap but you can leave this as is if it's easier to read

File: pages/selfCheckQuestions.tsx [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+            _id: `60e642d7e4a1ae34207a92a${count}`,
+            type: "multiple_choice",
+            question: "autem sunt eiusdolores nesciunt impedit?",
+            options: ["1", "2", "3"],
+            questionNumber: count,
+        };
+        const newList = list.slice(0);
+        const newCount = count + 1;
+        newList.splice(index + 1, 0, newQuestion);
+        setCount(newCount);
```
Comment (nimwijetunga) [inline pages/selfCheckQuestions.tsx:146]:
```suggestion
        setCount(count + 1);
```

File: pages/selfCheckQuestions.tsx [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+        setList(newList);
+    }
+
+    function removeOneQuestion(id) {
+        const newList = list.filter((item) => item._id !== id);
+        setList(newList);
+    }
+    function removeAllQuestions() {
+        const newList = [];
+        setList(newList);
```
Comment (nimwijetunga) [inline pages/selfCheckQuestions.tsx:156]:
```suggestion
        setList([]);
```

File: pages/selfCheckQuestions.tsx [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+        const newList = list.slice(0);
+        const newCount = count + 1;
+        newList.splice(index + 1, 0, newQuestion);
+        setCount(newCount);
+        setList(newList);
+    }
+
+    function removeOneQuestion(id) {
+        const newList = list.filter((item) => item._id !== id);
+        setList(newList);
```
Comment (nimwijetunga) [inline pages/selfCheckQuestions.tsx:152]:
```suggestion
        setList(list.filter((item) => item._id !== id));
```

File: pages/selfCheckQuestions.tsx (hunk 4 above)
Comment (nimwijetunga) [reply pages/selfCheckQuestions.tsx:193]: We should add a TODO here so it's easier to find when searching through the codebase/if someone else picks up the ticket

File: src/components/selfCheckQuestion/index.tsx [hunk 17]
```diff
@@ -0,0 +1,450 @@
...
+                                    {choice.label}
+                                </option>
+                            ))}
+                        </Select>
+
+                        {questionIndex != 0 && (
+                            <IconButton
+                                onClick={() => onMoveUpQuestion(questionIndex)}
+                                ml={
+                                    questionIndex + 1 == selfCheckQuestionSize
```
Comment (nimwijetunga) [inline src/components/selfCheckQuestion/index.tsx:151]: whats the purpose of this?

File: src/components/selfCheckQuestion/index.tsx [hunk 18]
```diff
@@ -0,0 +1,450 @@
...
+                                                    size="15px"
+                                                    borderWidth="2px"
+                                                    borderColor="black"
+                                                    borderRadius="sm"
+                                                ></Square>
+                                            }
+                                        />
+                                        <Input
+                                            variant="flushed"
+                                            placeholder={`Option ${index + 1}`}
```
Comment (nimwijetunga) [inline src/components/selfCheckQuestion/index.tsx:235]: onChange?

File: src/components/selfCheckQuestion/index.tsx [hunk 19]
```diff
@@ -0,0 +1,450 @@
...
+                                    variant="ghost"
+                                    onClick={() => onAddOption(questionId)}
+                                >
+                                    +Add Option
+                                </Button>
+                            </Stack>
+                        </Flex>
+                    )}
+                    {type == "short_answer" && (
+                        <Textarea
```
Comment (nimwijetunga) [inline src/components/selfCheckQuestion/index.tsx:263]: should we impose a word limit here since it's short answer

Comment (nimwijetunga) [reply src/components/selfCheckQuestion/index.tsx:263]: Also we need an onchange here

File: src/components/selfCheckQuestion/index.tsx [hunk 20]
```diff
@@ -0,0 +1,450 @@
...
+                            height={10}
+                            maxHeight={300}
+                            resize={"vertical"}
+                            mr={3}
+                            placeholder={"Short answer text"}
+                        />
+                    )}
+                    {type == "long_answer" && (
+                        <Textarea
+                            height={10}
```
Comment (nimwijetunga) [inline src/components/selfCheckQuestion/index.tsx:273]: `onChange`?

File: src/components/selfCheckQuestion/index.tsx [hunk 21]
```diff
@@ -0,0 +1,450 @@
...
+                                                fontWeight="500"
+                                                alignSelf="center"
+                                                mr={2}
+                                            >
+                                                {index + 1}
+                                            </Heading>
+                                        )}
+                                        <Input
+                                            index={index}
+                                            variant="flushed"
```
Comment (nimwijetunga) [inline src/components/selfCheckQuestion/index.tsx:340]: `onChange`

File: pages/selfCheckQuestions.tsx (hunk 1 above)
Comment (ChinemeremChigbo) [reply pages/selfCheckQuestions.tsx:20]: > The UI looks great! I left a few comments, will go through more of the PR soon
> 
> Small bug(?) but the console log on save does not show me the correct user input for any title/question values I added. I can see the number of questions and type updating, but the actual content remains the same always

This should be fixed now

**Review by nimwijetunga — CHANGES_REQUESTED** [review]
Great work on the UI it looks really good! Did a first pass review, main thing was that you're missing a few `onChange` that result in the data not getting updated when you hit save

**Review by ChinemeremChigbo — COMMENTED** [review]
> Great work on the UI it looks really good! Did a first pass review, main thing was that you're missing a few `onChange` that result in the data not getting updated when you hit save

Should be fixed now!

File: pages/selfCheckQuestions.tsx (hunk 10 above)
Comment (ChinemeremChigbo) [reply pages/selfCheckQuestions.tsx]: It's mostly for the db, it tells us whether to store our values as an float or alphanumerically

File: pages/selfCheckQuestions.tsx [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+                                removeAllQuestions();
+                            }}
+                            w={100}
+                            colorScheme="red"
+                        >
+                            Discard
+                        </Button>
+                    </ModalFooter>
+                </ModalContent>
+            </Modal>
```
Comment (mkanoria) [inline pages/selfCheckQuestions.tsx:223]: If you want to check out @daniel-sm-yu's work on the global Modal component (#27) , it might be useful to modify this into using that. I know it does essentially the same work so I'll leave this as your decision

File: src/components/selfCheckQuestion/index.tsx [hunk 23]
```diff
@@ -0,0 +1,852 @@
...
+    ModalContent,
+    ModalFooter,
+    ModalBody,
+    ModalCloseButton,
+} from "@chakra-ui/react";
+
+import { ArrowUpIcon, ArrowDownIcon } from "@chakra-ui/icons";
+
+//Self check question card component
+export const SelfCheckQuestionCard: React.FC = ({
```
Comment (mkanoria) [inline src/components/selfCheckQuestion/index.tsx:33]: I would recommend adding the props that this component uses as an Interface - would just make finding bugs easier. Should be a quick fix, here's an example for reference https://github.com/uwblueprint/headsupguys/pull/27/files#diff-a77615b501764f7894a235e85ccffdd4513bbabff29289cd6fda94ee0f6df228R14

File: src/components/selfCheckQuestion/index.tsx (hunk 17 above)
Comment (mkanoria) [reply src/components/selfCheckQuestion/index.tsx:162]: I'm assuming this is to figure out whether there's a 'move question down' button needed or not

File: src/components/selfCheckQuestion/index.tsx (hunk 19 above)
Comment (mkanoria) [reply src/components/selfCheckQuestion/index.tsx:683]: > should we impose a word limit here since it's short answer

this would be good idea - let's discuss with the PM/designers

as for the onchange i don't know if we really need that - since the textarea is more so for the users to fill out right? there's no default values afaik, we should also confirm this with the product team

**Comment by mkanoria** [comment]
Left 2 small comments and there's a few things I want to confirm with the PM's about before merging -- the new changes look  great and I really like the corresponding values addition!

File: src/components/selfCheckQuestion/index.tsx (hunk 17 above)
Comment (ChinemeremChigbo) [reply src/components/selfCheckQuestion/index.tsx:162]: On the very last question, there should only be the move up button. As a result the move down button is not shown, as a result the left margin must be increased to make sure everything is still lined up with the other questions

File: src/components/selfCheckQuestion/index.tsx (hunk 19 above)
Comment (ChinemeremChigbo) [reply src/components/selfCheckQuestion/index.tsx:683]: I think the word limit should only be on the client side, in fact, after talking to the product team I don't think the short answer on the admin side should even be an editable textbox, and is rather just there to show what the textbox would look like for the end user

**Review by nimwijetunga — CHANGES_REQUESTED** [review]
Changes seem to be registering now! One thing to take a look at is that the first two questions always seem to have a question number of 1: 

image

I don't know if this is b/c u initially set the react state to be question 1 but it should probably be looked into before merging

**Comment by ChinemeremChigbo** [comment]
> Changes seem to be registering now! One thing to take a look at is that the first two questions always seem to have a question number of 1:
> 
> image
> 
> I don't know if this is b/c u initially set the react state to be question 1 but it should probably be looked into before merging

should be fixed!

File: pages/selfCheckQuestions.tsx [hunk 24]
```diff
@@ -64,7 +64,7 @@ const Home: React.FC = () => {
     };
     const changeOptionInput = (id, index, target, optionOrValue) => {
         const newList = questionList.slice(0);
-        const changeIndex = optionOrValue == "option" ? 0 : 1;
+        const changeIndex = optionOrValue == "option" ? 0 : 1;0
```
Comment (nimwijetunga) [inline pages/selfCheckQuestions.tsx:67]: is this chance intentional?

Comment (nimwijetunga) [inline pages/selfCheckQuestions.tsx:67]: is this change intentional?

**Review by nimwijetunga — APPROVED** [review]
lgtm

**Review by mkanoria — APPROVED** [review]
LGTM!

**Comment by mkanoria** [comment]
@ChinemeremChigbo from our chat I remember this PR being complete (- the merge conflict) which is now resolved. So merging this in. Great job again!
